src/cmd/devdraw/wayland.c

```diff
@@ -422,27 +427,59 @@ static const struct xdg_toplevel_listener xdg_toplevel_listener = {
 	.close = xdg_toplevel_close,
 };
 
-static const struct wl_callback_listener wl_callback_listener;
+static const struct wl_callback_listener wl_callback_key_repeat_listener;
+
+static void wl_callback_key_repeat(void *data, struct wl_callback *wl_callback, uint32_t time) {
+	int dt = 0;
+	int repetitions = 0;
+	int repeat_rune = 0;
 
-// The callback is called per-frame.
-// It is currently only used to implement key repeat.
-static void wl_callback_done(void *data, struct wl_callback *wl_callback, uint32_t time) {
 	Client* c = data;
 	WaylandClient *wl = (WaylandClient*) c->view;
+	wl_callback_destroy(wl_callback);
+
 	qlock(&wayland_lock);
+	dt = time - wl->repeat_start_ms;
 
-	// Request another callback for the next frame.
-	// TODO: Only request a callback if we are still repeating a key.
-	wl_callback_destroy(wl_callback);
+	if (!wl->repeat_interval_ms || !wl->repeat_rune)
+		goto done;
```

Comment:
I think the rest of the file uses { } ever around single-line if blocks. Mind adding those to keep it consistent?

Reply:
done!